res/controllers/Reloop-Mixage.scripts.js

```diff
@@ -0,0 +1,358 @@
+// Name: Reloop Mixage
+// Author: Bim Overbohm
+// Version: 1.0.4 needs Mixxx 2.1+
+
+var Mixage = {};
+
+// ----- User-configurable settings -----
+Mixage.scratchByWheelTouch = false; // turn on scratching by touching the wheel instead of having to press the disc button
+
+// ----- Internal variables (don't touch) -----
+Mixage.vuMeterConnection = [];
+Mixage.libraryHideTimer = 0;
+Mixage.libraryReducedHideTimeout = 500;
+Mixage.libraryHideTimeout = 4000;
+Mixage.libraryRemainingTime = 0;
+Mixage.scratchPressed = false;
+Mixage.scrollPressed = false;
+Mixage.beatMovePressed = false;
+Mixage.effectRackSelected = [[true, false], [true, false]]; // if effect rack 1/2 is selected for channel 1/2
+Mixage.effectRackEnabled = [false, false]; // if effect rack 1/2 is enabled for channel 1/2
+
+Mixage.init = function(/*id, debugging*/) {
+    Mixage.connectControlsToFunctions("[Channel1]");
+    Mixage.connectControlsToFunctions("[Channel2]");
+    // all button LEDs off
+    for (var i = 0; i < 255; i++) {
+        midi.sendShortMsg(0x90, i, 0);
+    }
+    // start timers for updating the VU meters
+    Mixage.vuMeterConnection[0] = engine.makeConnection("[Channel1]", "VuMeter", function(val) {
+        midi.sendShortMsg(0x90, 29, val * 7);
+    });
+    Mixage.vuMeterConnection[1] = engine.makeConnection("[Channel2]", "VuMeter", function(val) {
+        midi.sendShortMsg(0x90, 30, val * 7);
+    });
+    // disable scrating on both decks
+    engine.scratchDisable(1);
+    engine.scratchDisable(2);
```

Comment:
The default on this controller (from the manual and demonstration videos) is that scratching is disabled until the "disc" button is toggled. I almost have this working... I want to disable scratching while the deck is stopped and while the deck is playing. The above two lines only disable scratching while the deck is playing... Any ideas?

Reply:
See this [video](https://www.youtube.com/watch?v=mYIvkGWRC4E) and the [manual](https://www.reloop.com/media/catalog/product/pdf/2/2/3/223618_Reloop_IM.pdf).

Reply:
Thank you, the video was very insightful. The integration should be very straightforward then. I'm afraid you're misunderstanding the scratching API a little bit. If you call `engine.scratchDisable` or `engine.scratchEnable`, you are signalling that you are touching the top of the jogwheel. That is completely unrelated to the `jog` control used for the tempo/pitch-bend when only touching the side.

So to implement the controllers behavior in mixxx: if no mode is selected any midi message for the platter touch and for the platter turn can simply be ignored using an early return in the JS handler functions. 
If the scratch mode is selected, touching the top should simply call `engine.scratchEnable(...)` while releasing it should call `engine.scratchDisable(...)`. In scratch mode the handler for the platter turn should check whether the deck is scratching (`engine.isScratching()` iirc) and then should either call `engine.scratchTick` or the `jog` CO (depending on whether the top was touched or not). You can check the implementation of `components.JogWheelBasic` for an example. 
For the Search mode, you can experiment a little. I think setting the `playPosition` CO directly would probably yield the best results. You're also free to decide whether the top of the platter should be taken into account. The video didn't mention any difference. It would probably simpler to ignore it. 

Does that make sense?

Reply:
Thanks for the explanation! So engine.isScratching() is just scratching vs. pitch bend. I need an extra variable then...
playPosition looks very odd. How on earth do I use that?! Atm I set different values in scratchEnable() depending on the toggled mode (128 vs. 620 ticks per revolution). Would that work too?

Reply:
> Thanks for the explanation! So engine.isScratching() is just scratching vs. pitch bend. I need an extra variable then...

Yes, you should add an extra variable per deck that tracks the "platter-mode" its currently in. 

> playposition = playposition + fudge_factor * control_value_diff?

yeah, something like that. Scratching with a higher factor (or less ticks, should be around the same), would work too. Just experiment with what sounds better and feels more responsive. I think playposition would be better though because that is less likely to break in case we improve the scratching code in the future.

Reply:
I've used playposition now. It is good in that it mutes the track when scrolling.